internal/controllers/apis/restarton/restarton.go

```diff
@@ -235,3 +247,18 @@ func FilesChanged(restartOn *v1alpha1.RestartOnSpec, fileWatches map[string]*v1a
 	}
 	return sliceutils.DedupedAndSorted(filesChanged)
 }
+
+// RegisterWatches ensures that reconciliation happens on changes to objects referenced by RestartOnSpec/StartOnSpec.
+//
+// It should be called by a controllers CreateBuilder() method.
+//
+// Controllers MUST also call ExtractKeysForIndexer as part of their indexer initialization.
+func RegisterWatches(builder *builder.Builder, indexer *indexer.Indexer) {
+	for _, t := range restartOnTypes {
+		// this is arguably overly defensive, but a copy of the type object stub is made
+		// to avoid sharing references of it across different reconcilers
+		obj := t.DeepCopyObject().(client.Object)
+		builder.Watches(&source.Kind{Type: obj},
+			handler.EnqueueRequestsFromMapFunc(indexer.Enqueue))
```

Comment:
feels like we're missing some underlying abstraction. 

This implicitly relies on RestartOn already being registered with the indexer.

I wonder if the indexer should have some sort of plugin system where you're registering both Watches and Key indexers?

Reply:
Yeah, if the `indexer.Indexer` accepted multiple `KeyFunc`s and then took the union of them all, we could do this in one shot.

You'll still need an "extractor" func to go from `client.Object` -> `(*v1alpha1.RestartOnSpec, *v1alpha1.StartOnSpec)` but that's not the worst 🤷 

I'll play around with it briefly and see if I can open another PR that combines them